storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -88,85 +488,214 @@ type ReconcileVolume struct {
-			// Return and don't requeue
-			// TODO Remove the finalizer from the generated PV, if exists
-			return reconcile.Result{}, nil
+			// Volume not found:
+			// => all the finalizers have been removed & Volume has been deleted
+			// => there is nothing left to do
+			reqLogger.Info("volume already deleted: nothing to do")
+			return endReconciliation()
 		}
-		// Error reading the object - requeue the request.
-		return reconcile.Result{}, err
+		reqLogger.Error(err, "cannot read Volume: requeue")
+		return delayedRequeue(err)
 	}
-
-	result, err := r.salt.TestPing()
-	if err != nil {
-		reqLogger.Error(err, "test.ping failed")
-		return reconcile.Result{}, err
+	// This check should be done by an Admission Controller…
+	if err := volume.IsValid(); err != nil {
+		return r.setFailedVolumeStatus(
+			ctx, volume, storagev1alpha1.InternalError,
+			"invalid volume: %s", err.Error(),
+		)
 	}
-	for hostname := range result {
+	// Check if the volume is marked for deletion (i.e., deletion tsamp is set).
+	if !volume.GetDeletionTimestamp().IsZero() {
+		return r.finalizeVolume(ctx, request, volume)
+	}
+	// Skip volume stuck waiting for deletion or a manual fix.
+	if volume.IsInUnrecoverableFailedState() {
 		reqLogger.Info(
-			"ping", "hostname", hostname, "result", result[hostname].(bool),
+			"volume stuck in error state: do nothing",
+			"Error.Code", volume.Status.ErrorCode,
+			"Error.Message", volume.Status.ErrorMessage,
 		)
+		return endReconciliation()
 	}
-
-	pv := newPersistentVolumeForCR(instance)
-
-	// Set Volume instance as the owner and controller
-	if err := controllerutil.SetControllerReference(instance, pv, r.scheme); err != nil {
-		return reconcile.Result{}, err
+	// Check if a PV already exists for this volume.
+	pv, err := r.getPersistentVolume(ctx, volume.Name)
+	if err != nil {
+		reqLogger.Error(
+			err, "cannot read PersistentVolume: requeue",
+			"PersistentVolume.Name", volume.Name,
+		)
+		return delayedRequeue(err)
 	}
+	// PV doesn't exist: deploy the volume to create it.
+	if pv == nil {
+		return r.deployVolume(ctx, request, volume)
+	}
+	// Else, check its health.
+	if pv.Status.Phase == corev1.VolumeFailed {
+		_, err := r.setFailedVolumeStatus(
+			ctx, volume, storagev1alpha1.UnavailableError,
+			"backing PersistentVolume is in a failed state (%s): %s",
+			pv.Status.Reason, pv.Status.Message,
+		)
+		return delayedRequeue(err)
+	}
+	if _, err = r.setAvailableVolumeStatus(ctx, volume); err != nil {
+		return delayedRequeue(err)
+	}
+	reqLogger.Info("backing PersistentVolume is healthy")
+	return endReconciliation()
+}
+
+// Deploy a volume (i.e prepare the storage and create a PV).
+func (self *ReconcileVolume) deployVolume(
+	ctx context.Context,
+	request reconcile.Request,
+	volume *storagev1alpha1.Volume,
+) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Name", request.Name)
+	nodeName := string(volume.Spec.NodeName)
```

Comment:
Adding this to the logger context could be useful.

Reply:
Indeed. Done